### PR TITLE
Add getting contact by field

### PR DESCRIPTION
**Pull request by rudigiesler**
We want to be able to get a contact by specifying a specific field, for example `msisdn`. The queries should have the form `?q='msisdn=+1234'`

### CODE REVIEW

**Comment by SteveBarnett** [comment]
@rudigiesler Do we match for a missing / present `+`?
I mean if someone searches for `+123`, would they see `123`. And if they searched for `+123` would they  see `123` (i guess yes).

**Comment by rudigiesler** [comment]
@SteveBarnett Not at the moment. It's an exact match for now. We could do that, but it would make things a little more complicated (we would have to have different things for different field types, eg. msisdn we would have to look for `+`, but twitter_id we would have to look for `@`)

**Comment by rudigiesler** [comment]
I think the tests are failing due to the riak version, checking up on this.

**Comment by SteveBarnett** [comment]
@rudigiesler How about a shorter / simpler version: strip leading `+`s from msisdns and leading `@`s from twitter ids (strip from the thing that is thrown into the query, not from the field value itself, I mean)? Or do we do that already?

**Comment by jerith** [comment]
@SteveBarnett This does a Riak index query under the hood, so we can only look up exact values. We could manipulate the query string before we search, but we don't necessarily have `+` at the front of all MSISDNs, etc.

File: go_contacts/backends/contacts.py [hunk 1]
```diff
@@ -91,6 +94,33 @@ def all_keys(self):
...
+
+    @inlineCallbacks
+    def _get_contacts_by_query(self, query):
+        try:
+            [field, value] = query.split('=')
+        except ValueError:
+            raise CollectionUsageError(
+                "Query must be of the form 'field=value'")
+        valid_keys = self.delivery_classes.keys()
+        if field not in valid_keys:
```
Comment (hodgestar) [inline go_contacts/backends/contacts.py]: I think we could combine these two lines into ``if field not in self.delivery_classes``.

Comment (hodgestar) [inline go_contacts/backends/contacts.py]: And just use ``self.delivery_classes.keys()`` in the error message (perhaps sorted?).

**Comment by hodgestar** [comment]
@SteveBarnett @jerith @rudigiesler ``contact_field_for_addr`` which is called by ``contact_for_addr`` already ensures that ``msisdn`` values have a ``+`` on the left and does some sanity adjustments for Gtalk addresses.

**Comment by hodgestar** [comment]
Slightly radical suggestion -- long term we're trying to get rid of ``DELIVERY_CLASS`` in favour of ``address_type``s and the ``delivery_class`` isn't really relevant here -- we just want to look up a contact given a value for one of it's fields. I propose that we tweak the methods on the contact model in Vumi Go so that there is an underlying method ``contact_for_addr_field(field, value)`` that is called by the current ``contact_for_addr``. We can then call that method here too and just keep a list of field names (maybe the field list could be a new tuple attribute on the ``Contact`` model).

@rudi Thoughts?

**Comment by rudigiesler** [comment]
@hodgestar: It is something we can do, but it will add a bunch of time to this already unplanned task. I'll leave the decision up to you as whether it is worth the extra time or not.

**Comment by hodgestar** [comment]
@rudi I think let's do this -- it's a little extra work, but it helps kill delivery classes, which is a worthy goal. (copied from Slack for the record).

**Comment by rudigiesler** [comment]
@hodgestar This is now using the new vumi-go things. Ready for review.

File: go_contacts/backends/contacts.py [hunk 2]
```diff
@@ -91,6 +92,28 @@ def all_keys(self):
...
+        except ValueError:
+            raise CollectionUsageError(
+                "Query must be of the form 'field=value'")
+        if field not in Contact.ADDRESS_FIELDS:
+            raise CollectionUsageError(
+                "Query field must be one of: %s" %
+                sorted(Contact.ADDRESS_FIELDS))
+
+        try:
+            value = normalize_addr(field, value)
```
Comment (hodgestar) [inline go_contacts/backends/contacts.py]: Can the call to ``normalize_addr`` be moved outside the ``try:`` block?

**Comment by hodgestar** [comment]
:+1: as soon as Travis is happy.

**Pull request merged**